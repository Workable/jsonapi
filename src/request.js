import request from "@workablehr/request";
import deserialize from "./deserialize";

const requestJsonapi = (...args) => {
  const requestPromise = request(...args);
  const deserializePromise = requestPromise.then(deserialize).catch(error => {
    throw deserialize(error.response);
  });
  deserializePromise.abort = requestPromise.abort;
  return deserializePromise;
};

requestJsonapi.post = (url, data, params) =>
  requestJsonapi(url, {
    ...params,
    method: "POST",
    ...(data && { body: JSON.stringify(data) })
  });

requestJsonapi.put = (url, data, params) =>
  requestJsonapi(url, {
    ...params,
    method: "PUT",
    ...(data && { body: JSON.stringify(data) })
  });

export default requestJsonapi;
