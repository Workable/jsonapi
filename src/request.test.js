import request from "./request";
import resource from "./fixtures/resource";
import errors from "./fixtures/errors";

let timeoutId;
let abortSpy;
let realAbortController;

beforeEach(() => {
  realAbortController = global.AbortController;

  global.AbortController = function() {
    return {
      signal: {},
      abort: () => {
        clearTimeout(timeoutId);
        abortSpy();
      }
    };
  };

  clearTimeout(timeoutId);
  abortSpy = jest.fn();
});

afterEach(() => {
  global.AbortController = realAbortController;
});

it("submits a request", async () => {
  fetch.mockResponseOnce(JSON.stringify(resource));
  const response = await request("mockUrl");
  expect(response).toMatchSnapshot();
  expect(abortSpy).toBeCalledTimes(0);
});

it("aborts a request", async () => {
  jest.useFakeTimers();
  fetch.mockResponseOnce(
    () =>
      new Promise(
        resolve =>
          (timeoutId = setTimeout(
            () => resolve({ body: JSON.stringify(resource) }),
            0
          ))
      )
  );
  const successSpy = jest.fn();
  const promise = request("mockUrl");
  promise.then(successSpy);
  promise.abort();

  jest.advanceTimersByTime(10);
  expect(successSpy).not.toBeCalled();
  expect(abortSpy).toBeCalledTimes(1);
});

it("returns the jsonapi error", async () => {
  // should parse error content
  fetch.mockResponseOnce(new Error(JSON.stringify(errors)), { status: 422 });
  let error;
  try {
    await request("mockUrl");
  } catch (e) {
    error = e;
  }
  expect(error.status).toBe(422);
});
