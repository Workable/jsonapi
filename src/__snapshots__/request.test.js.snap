// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submits a request 1`] = `
Object {
  "account": Array [
    Object {
      "id": "6969",
      "name": "My Account",
      "type": "account",
    },
  ],
  "job": Object {
    "accountId": "6969",
    "benefits": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor.",
    "city": "Anchorage",
    "countryCode": "US",
    "countryName": "United States",
    "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt.",
    "id": "210136031",
    "industry": "Marketing and Advertising",
    "keywords": Array [
      "general management",
    ],
    "questionIds": Array [
      "123",
      "234",
    ],
    "requirements": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. <ul><li>Aenean commodo ligula eget dolor?</li></ul>",
    "stateCode": "AK",
    "subregion": "Alaska",
    "title": "Head of Security",
    "type": "job",
  },
  "meta": Object {
    "ctx": "22563a70-8bd2-4719-9d1b-7c610820e8bc",
  },
  "question": Array [
    Object {
      "body": "some question",
      "id": "123",
      "type": "question",
    },
    Object {
      "body": "some question",
      "id": "234",
      "type": "question",
    },
  ],
}
`;
